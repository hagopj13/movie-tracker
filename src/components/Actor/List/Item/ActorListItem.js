// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import ImageIcon from '@material-ui/icons/Image';

import configSelectors from 'store/config/config.selectors';
import type { Actor } from 'types';

type Props = {
  actor: Actor,
  fullProfilePath: string,
};

const useStyles = makeStyles((theme) => ({
  root: {
    width: 155,
    minWidth: 155,
  },
  cardContent: {
    padding: theme.spacing(2),
    '&:last-child': {
      paddingBottom: theme.spacing(2),
    },
  },
  profileMedia: {
    width: 155,
    height: 235,
  },
  brokenMediaContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: theme.palette.grey[200],
    height: 235,
  },
}));

export const ActorListItem = (props: Props) => {
  const { actor, fullProfilePath } = props;

  const classes = useStyles();

  const isImageFound = actor.profilePath && fullProfilePath;

  const renderMedia = () => {
    return isImageFound ? (
      <CardMedia className={classes.profileMedia} image={fullProfilePath} />
    ) : (
      <div className={classes.brokenMediaContainer}>
        <ImageIcon fontSize="large" />
      </div>
    );
  };

  return (
    <Card className={classes.root} elevation={3}>
      {renderMedia()}
      <CardContent className={classes.cardContent}>
        <Typography variant="h6">{actor.name}</Typography>
        <Typography>{actor.character}</Typography>
      </CardContent>
    </Card>
  );
};

const mapStateToProps = createStructuredSelector({
  fullProfilePath: (state, ownProps: Props) =>
    configSelectors.createProfileFullPathSelector(ownProps.actor.profilePath)(state),
});

export default connect(mapStateToProps)(ActorListItem);
