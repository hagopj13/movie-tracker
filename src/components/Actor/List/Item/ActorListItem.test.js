import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import ImageIcon from '@material-ui/icons/Image';

import { ActorListItem } from './ActorListItem';

describe('ActorListItem component', () => {
  let shallow;
  let actor;
  let fullProfilePath;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    actor = {
      id: 'someId',
      name: 'Some name',
      character: 'Some character',
      profilePath: 'pathToProfileImage.png',
    };
    fullProfilePath = 'fullPathToProfileImage.png';
    const mockProps = {
      actor,
      fullProfilePath,
    };
    wrapper = shallow(<ActorListItem {...mockProps} />);
  });

  it('should render the ActorListItem component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render CardMedia if fullProfilePath is defined', () => {
    expect(wrapper.exists(CardMedia)).toBe(true);
    expect(wrapper.find(CardMedia).prop('image')).toBe(fullProfilePath);
  });

  it('should render a div containing ImageIcon as media if fullProfilePath is not defined', () => {
    const mockProps = {
      actor,
      fullProfilePath: null,
    };
    const newWrapper = shallow(<ActorListItem {...mockProps} />);
    expect(newWrapper.exists(CardMedia)).toBe(false);
    expect(newWrapper.exists(ImageIcon)).toBe(true);
  });

  it('should render actor name in CardContent', () => {
    expect(wrapper.find(CardContent).childAt(0).text()).toBe(actor.name);
  });

  it('should render character name in CardContent', () => {
    expect(wrapper.find(CardContent).childAt(1).text()).toBe(actor.character);
  });
});
