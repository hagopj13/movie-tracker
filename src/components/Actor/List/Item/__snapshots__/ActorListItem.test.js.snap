// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActorListItem component should render the ActorListItem component correctly 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-root-1"
  elevation={3}
>
  <WithStyles(ForwardRef(CardMedia))
    className="makeStyles-profileMedia-3"
    image="fullPathToProfileImage.png"
  />
  <WithStyles(ForwardRef(CardContent))
    className="makeStyles-cardContent-2"
  >
    <WithStyles(ForwardRef(Typography))
      variant="h6"
    >
      Some name
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))>
      Some character
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(CardContent))>
</WithStyles(ForwardRef(Card))>
`;
